### PR TITLE
Rest API

### PR DESCRIPTION
**Pull request by brittainhard**

### CODE REVIEW

**Comment by brittainhard** [comment]
@ahmadia this is ready for review

**Comment by ahmadia** [comment]
Okay, testing it out.

File: source/memex/common_settings.py [hunk 1]
```diff
@@ -159,3 +160,12 @@
     sys.stderr.write("[%d]: The following applications are ready %s\n" % (os.getpid(), str(READY_EXTERNAL_APPS)))
 else:
     sys.stderr.write("[%d]: Supervisord not running or no applications are running\n" % (os.getpid()))
+
+REST_FRAMEWORK = {
+    # Use Django's standard `django.contrib.auth` permissions,
```
Comment (ahmadia) [inline source/memex/common_settings.py]: These comments are confusing.  Access is unrestricted with this field.

Comment (brittainhard) [inline source/memex/common_settings.py]: Yeah that's just copy-pasted. Forgot to remove it.

File: source/apps/crawl_space/models.py [hunk 2]
```diff
@@ -167,53 +161,36 @@ def get_solr_url(self):
...
+        return super(Crawl, self).save(*args, **kwargs)
 
-
-            # Continue saving as normal
-
-        self.slug = slugify(unicode(self.name))
-        super(Crawl, self).save(*args, **kwargs)
-
+    # TODO:
+    # This is redundant. URL property is better.
```
Comment (ahmadia) [inline source/apps/crawl_space/models.py]: You could just hide this method by renaming it to _get_absolute_url().  I'm okay with keeping it as its own method.

File: source/apps/crawl_space/models.py [hunk 3]
```diff
@@ -167,53 +161,36 @@ def get_solr_url(self):
...
-                ensure_exists(seed_dir)
-                dst = join(crawl_path, 'seeds/seeds')
-                shutil.move(self.seeds_list.path, dst)
-                self.seeds_list.name = seed_dir
-            else:
-                dst = join(crawl_path, 'seeds')
-                shutil.move(self.seeds_list.path, dst)
-                self.seeds_list.name = dst
-                # Create unique configs for every ache crawl.
+            # If the crawler is ache, copy the config. If config already exists,
```
Comment (ahmadia) [inline source/apps/crawl_space/models.py]: Thanks for adding these notes.

File: source/apps/crawl_space/models.py [hunk 4]
```diff
@@ -167,53 +161,36 @@ def get_solr_url(self):
...
-        #    crawl directory.
         if self.pk is None:
-            # Need to save first to obtain the pk attribute.
             self.slug = slugify(unicode(self.name))
             self.location = os.path.join(resources_dir, "crawls", self.slug)
-            super(Crawl, self).save(*args, **kwargs)
-
-            # Ensure that the crawl path `resources/crawls/<crawl.pk>` exists
+            # TODO:
+            # Fix this function and its weird side effect. Without this line the
```
Comment (ahmadia) [inline source/apps/crawl_space/models.py]: What's the weird side effect?

File: source/apps/crawl_space/models.py [hunk 5]
```diff
@@ -69,23 +70,15 @@ def get_absolute_url(self):
...
-
     def save(self, *args, **kwargs):
-
         if self.pk is None:
-            super(CrawlModel, self).save(*args, **kwargs)
-
+            self.slug = slugify(self.name)
+            # TODO:
+            # Another weird call with a side effect that has to be fixed.
             model_path = self.ensure_model_path()
```
Comment (ahmadia) [inline source/apps/crawl_space/models.py]: I expect that "ensure_model_path" will make sure a directory/path has been created and contains appropriate content.  Is it doing something beyond this?

**Comment by ahmadia** [comment]
LGTM

File: source/apps/crawl_space/models.py (hunk 5 above)
Comment (brittainhard) [inline source/apps/crawl_space/models.py]: That's pretty much it. I left a TODO there because it's not exactly a pressing matter.

Comment (ahmadia) [inline source/apps/crawl_space/models.py]: Then I don't think that's a weird side effect, that's what the function should be doing.

**Comment by ahmadia** [comment]
Thanks Brittain, really nice work.

File: source/apps/crawl_space/models.py (hunk 5 above)
Comment (brittainhard) [inline source/apps/crawl_space/models.py]: You're right. I think the "weirdness" comes from the way the side effect is done, not that it has a side effect. I usually try to stay away from calling functions in other functions just for the purpose of causing a side effect, i.e.: 
```
ensure_exists(crawl_path)
```
Not how I would have done it.

**Comment by brittainhard** [comment]
Thanks @ahmadia

**Comment by chrismattmann** [comment]
Why are we doing our own REST API for this instead of using the Nutch one?
http://wiki.apache.org/nutch/Nutch_1.X_RESTAPI

I'm pretty confused here. Perhaps the reason is to isolate what crawler is being used and to do a new REST API based on that. Is that the reason? @brittainhard @ahmadia

**Comment by chrismattmann** [comment]
even if the reason is to define a Memex Explorer specific REST api, I strongly suggest we should be basing it off the Nutch one. It's fairly complete?

**Comment by brittainhard** [comment]
Hey @chrismattmann this rest API is completely separate from the the Nutch REST Api. The primary use of this service is to serialize the memex explorer database objects into JSON, so that they can be easily rendered with Backbonejs views.

**Comment by ahmadia** [comment]
@chrismattmann - In other words, the REST API in Memex Explorer enables the modal AJAX interfaces that we recently implemented.

**Comment by chrismattmann** [comment]
+1 thanks dudes @brittainhard @ahmadia

**Pull request merged**